File: tools/dashboard/parsing.py
"""Parsing functions for Spacedock YAML frontmatter and entity scanning.

Copied from skills/commission/bin/status (which has no .py extension and
cannot be imported). Keep in sync manually if the source changes.
"""

import glob
import os


def parse_frontmatter(filepath):
    """Extract YAML frontmatter fields from a markdown file."""
    fields = {}
    in_fm = False
    with open(filepath, 'r') as f:
        for line in f:
            line = line.rstrip('\n')
            if line == '---':
                if in_fm:
                    break
                in_fm = True
                continue
            if in_fm:
                if ':' in line:
                    key, _, val = line.partition(':')
                    key = key.strip()
                    val = val.strip()
                    if not line[0].isspace():
                        fields[key] = val
    return fields


def parse_stages_block(filepath):
    """Parse the stages block from README frontmatter."""
    lines = []
    in_fm = False
    with open(filepath, 'r') as f:
        for line in f:
            line = line.rstrip('\n')
            if line == '---':
                if in_fm:
                    break
                in_fm = True
                continue
            if in_fm:
                lines.append(line)

    stages_start = None
    for i, line in enumerate(lines):
        if line.rstrip() == 'stages:':
            stages_start = i
            break

    if stages_start is None:
        return None

    defaults = {}
    states = []
    i = stages_start + 1
    stages_indent = None
    while i < len(lines):
        line = lines[i]
        stripped = line.lstrip()
        if not stripped:
            i += 1
            continue
        indent = len(line) - len(stripped)
        if stages_indent is None:
            stages_indent = indent
        elif indent < stages_indent:
            break

        if indent == stages_indent:
            if stripped == 'defaults:':
                i += 1
                while i < len(lines):
                    dline = lines[i]
                    dstripped = dline.lstrip()
                    if not dstripped:
                        i += 1
                        continue
                    dindent = len(dline) - len(dstripped)
                    if dindent <= stages_indent:
                        break
                    if ':' in dstripped:
                        k, _, v = dstripped.partition(':')
                        defaults[k.strip()] = v.strip()
                    i += 1
                continue
            elif stripped == 'states:':
                i += 1
                current_state = None
                while i < len(lines):
                    sline = lines[i]
                    sstripped = sline.lstrip()
                    if not sstripped:
                        i += 1
                        continue
                    sindent = len(sline) - len(sstripped)
                    if sindent <= stages_indent:
                        break
                    if sstripped.startswith('- name:'):
                        _, _, name = sstripped.partition('- name:')
                        current_state = {'name': name.strip()}
                        states.append(current_state)
                    elif current_state is not None and ':' in sstripped and not sstripped.startswith('- '):
                        k, _, v = sstripped.partition(':')
                        current_state[k.strip()] = v.strip()
                    i += 1
                continue
        i += 1

    if not states:
        return None

    default_worktree = defaults.get('worktree', 'false').lower() == 'true'
    default_concurrency = int(defaults.get('concurrency', '2'))

    result = []
    for state in states:
        stage = {
            'name': state['name'],
            'worktree': state.get('worktree', str(default_worktree)).lower() == 'true',
            'concurrency': int(state.get('concurrency', str(default_concurrency))),
            'gate': state.get('gate', 'false').lower() == 'true',
            'terminal': state.get('terminal', 'false').lower() == 'true',
            'initial': state.get('initial', 'false').lower() == 'true',
        }
        result.append(stage)

    return result


def scan_entities(directory):
    """Scan a directory for .md entity files (excluding README.md)."""
    entities = []
    pattern = os.path.join(directory, '*.md')
    for filepath in sorted(glob.glob(pattern)):
        if os.path.basename(filepath) == 'README.md':
            continue
        slug = os.path.splitext(os.path.basename(filepath))[0]
        fields = parse_frontmatter(filepath)
        entity = {k: v for k, v in fields.items()}
        entity['slug'] = slug
        for key in ('id', 'status', 'title', 'score', 'source', 'worktree'):
            entity.setdefault(key, '')
        entities.append(entity)
    return entities

File: tools/dashboard/src/parsing.ts
import { readFileSync } from "node:fs";
import { join } from "node:path";
import type { FrontmatterFields, Stage, Entity } from "./types";

export function parseFrontmatter(filepath: string): FrontmatterFields {
  const text = readFileSync(filepath, "utf-8");
  const lines = text.split("\n");
  const fields: FrontmatterFields = {};
  let inFm = false;

  for (const rawLine of lines) {
    const line = rawLine.replace(/\r$/, "");
    if (line === "---") {
      if (inFm) break;
      inFm = true;
      continue;
    }
    if (inFm && line.includes(":")) {
      if (line.length > 0 && line[0] !== " " && line[0] !== "\t") {
        const idx = line.indexOf(":");
        const key = line.slice(0, idx).trim();
        const val = line.slice(idx + 1).trim();
        fields[key] = val;
      }
    }
  }
  return fields;
}

export function parseStagesBlock(filepath: string): Stage[] | null {
  const text = readFileSync(filepath, "utf-8");
  const allLines = text.split("\n").map((l) => l.replace(/\r$/, ""));

  // Extract frontmatter lines
  const lines: string[] = [];
  let inFm = false;
  for (const line of allLines) {
    if (line === "---") {
      if (inFm) break;
      inFm = true;
      continue;
    }
    if (inFm) lines.push(line);
  }

  // Find stages: line
  let stagesStart: number | null = null;
  for (let i = 0; i < lines.length; i++) {
    if (lines[i].trimEnd() === "stages:") {
      stagesStart = i;
      break;
    }
  }
  if (stagesStart === null) return null;

  const defaults: Record<string, string> = {};
  const states: Record<string, string>[] = [];
  let i = stagesStart + 1;
  let stagesIndent: number | null = null;

  while (i < lines.length) {
    const line = lines[i];
    const stripped = line.trimStart();
    if (!stripped) { i++; continue; }
    const indent = line.length - stripped.length;
    if (stagesIndent === null) {
      stagesIndent = indent;
    } else if (indent < stagesIndent) {
      break;
    }

    if (indent === stagesIndent) {
      if (stripped === "defaults:") {
        i++;
        while (i < lines.length) {
          const dline = lines[i];
          const dstripped = dline.trimStart();
          if (!dstripped) { i++; continue; }
          const dindent = dline.length - dstripped.length;
          if (dindent <= stagesIndent!) break;
          if (dstripped.includes(":")) {
            const idx = dstripped.indexOf(":");
            defaults[dstripped.slice(0, idx).trim()] = dstripped.slice(idx + 1).trim();
          }
          i++;
        }
        continue;
      } else if (stripped === "states:") {
        i++;
        let currentState: Record<string, string> | null = null;
        while (i < lines.length) {
          const sline = lines[i];
          const sstripped = sline.trimStart();
          if (!sstripped) { i++; continue; }
          const sindent = sline.length - sstripped.length;
          if (sindent <= stagesIndent!) break;
          if (sstripped.startsWith("- name:")) {
            const name = sstripped.slice("- name:".length).trim();
            currentState = { name };
            states.push(currentState);
          } else if (currentState !== null && sstripped.includes(":") && !sstripped.startsWith("- ")) {
            const idx = sstripped.indexOf(":");
            currentState[sstripped.slice(0, idx).trim()] = sstripped.slice(idx + 1).trim();
          }
          i++;
        }
        continue;
      }
    }
    i++;
  }

  if (states.length === 0) return null;

  const defaultWorktree = (defaults.worktree ?? "false").toLowerCase() === "true";
  const defaultConcurrency = parseInt(defaults.concurrency ?? "2", 10);

  return states.map((state) => ({
    name: state.name,
    worktree: (state.worktree ?? String(defaultWorktree)).toLowerCase() === "true",
    concurrency: parseInt(state.concurrency ?? String(defaultConcurrency), 10),
    gate: (state.gate ?? "false").toLowerCase() === "true",
    terminal: (state.terminal ?? "false").toLowerCase() === "true",
    initial: (state.initial ?? "false").toLowerCase() === "true",
    feedback_to: state["feedback-to"] ?? "",
    conditional: (state.conditional ?? "false").toLowerCase() === "true",
    model: state.model ?? "",
  }));
}

/**
 * Parse a depends-on frontmatter value like "[007, 016]" into number[].
 * Returns empty array if the value is empty/undefined.
 */
export function parseDependsOn(raw: string | undefined): number[] {
  if (!raw) return [];
  const matches = raw.match(/\d+/g);
  if (!matches) return [];
  return matches.map(Number);
}

export function scanEntities(directory: string): Entity[] {
  const glob = new Bun.Glob("*.md");
  const files = Array.from(glob.scanSync({ cwd: directory })).sort();
  const entities: Entity[] = [];

  for (const filename of files) {
    if (filename === "README.md") continue;
    const filepath = join(directory, filename);
    const slug = filename.replace(/\.md$/, "");
    const fields = parseFrontmatter(filepath);
    const entity: Entity = {
      ...fields,
      slug,
      path: filepath,
      id: fields.id ?? "",
      status: fields.status ?? "",
      title: fields.title ?? "",
      score: fields.score ?? "",
      source: fields.source ?? "",
      worktree: fields.worktree ?? "",
    };
    entities.push(entity);
  }
  return entities;
}
